libraries/chain/nft_evaluator.cpp

```diff
@@ -0,0 +1,183 @@
+   auto itr_nft_md = idx_nft_md.find(op.nft_metadata_id);
+   FC_ASSERT( itr_nft_md != idx_nft_md.end(), "NFT metadata not found" );
+   FC_ASSERT( itr_nft_md->owner == op.owner, "Only owner can modify NFT metadata" );
+
+   return void_result();
+} FC_CAPTURE_AND_RETHROW( (op) ) }
+
+void_result nft_metadata_update_evaluator::do_apply( const nft_metadata_update_operation& op )
+{ try {
+   db().modify(db().get(op.nft_metadata_id), [&] ( nft_metadata_object& obj ) {
+      if( op.name.valid() )
+         obj.name = *op.name;
+      if( op.symbol.valid() )
+         obj.symbol = *op.symbol;
+      if( op.base_uri.valid() )
+         obj.base_uri = *op.base_uri;
+   });
+   return void_result();
+} FC_CAPTURE_AND_RETHROW( (op) ) }
+
+
+void_result nft_mint_evaluator::do_evaluate( const nft_mint_operation& op )
+{ try {
+   const auto& idx_nft_md = db().get_index_type<nft_metadata_index>().indices().get<by_id>();
+   auto itr_nft_md = idx_nft_md.find(*op.nft_metadata_id);
+   FC_ASSERT( itr_nft_md != idx_nft_md.end(), "NFT metadata not found" );
+   FC_ASSERT( itr_nft_md->owner == op.payer, "Only metadata owner can mint NFT" );
+
+   return void_result();
+} FC_CAPTURE_AND_RETHROW( (op) ) }
+
+object_id_type nft_mint_evaluator::do_apply( const nft_mint_operation& op )
+{ try {
+   const auto& new_nft_object = db().create<nft_object>( [&]( nft_object& obj ){
+      obj.nft_metadata_id = *op.nft_metadata_id;
+      obj.owner = op.owner;
+      obj.approved = op.approved;
+      obj.approved_operators = op.approved_operators;
+   });
+   return new_nft_object.id;
+} FC_CAPTURE_AND_RETHROW( (op) ) }
+
+
+void_result nft_safe_transfer_from_evaluator::do_evaluate( const nft_safe_transfer_from_operation& op )
+{ try {
+   const auto& idx_nft = db().get_index_type<nft_index>().indices().get<by_id>();
+   const auto& idx_acc = db().get_index_type<account_index>().indices().get<by_id>();
+
+   auto itr_nft = idx_nft.find(op.token_id);
+   FC_ASSERT( itr_nft != idx_nft.end(), "NFT does not exists" );
+
+   auto itr_operator = idx_acc.find(op.operator_);
+   FC_ASSERT( itr_operator != idx_acc.end(), "Operator account does not exists" );
+
+   auto itr_owner = idx_acc.find(itr_nft->owner);
+   FC_ASSERT( itr_owner != idx_acc.end(), "Owner account does not exists" );
+
+   auto itr_from = idx_acc.find(op.from);
+   FC_ASSERT( itr_from != idx_acc.end(), "Sender account does not exists" );
+   FC_ASSERT( itr_from->id == itr_owner->id, "Sender account is not owner of this NFT" );
+
+   auto itr_to = idx_acc.find(op.to);
+   FC_ASSERT( itr_to != idx_acc.end(), "Receiver account does not exists" );
+
+   auto itr_approved_op = std::find(itr_nft->approved_operators.begin(), itr_nft->approved_operators.end(), op.operator_);
+   FC_ASSERT( (itr_nft->owner == itr_owner->id) || (itr_nft->approved == itr_operator->id) || (itr_approved_op != itr_nft->approved_operators.end()), "Operator is not NFT owner or approved operator" );
```

Comment:
I think this owner check might allow anyone to transfer the NFT.

Please check if the merged code on marketplace branch is fine. 

https://github.com/peerplays-network/peerplays/pull/367/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.hpp&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-08317125078835b86641215a88c0e6f6R97